docs/adr/0023-paypal-support.md

```diff
@@ -0,0 +1,102 @@
+
+PayPal has a feature called reference transactions that allow FxA to get a customers authorization to charge their PayPal account. This authorization allows for later payments to be charged against the PayPal account. We will track this reference ID in Stripe metadata for the customer and mark the subscription as `send_invoice`.
+
+FxA will implement additional logic to listen for invoice creation webhooks and switch their `auto_advance` off if the subscription has metadata indicating it should be paid via a PayPal reference ID. We will then check the invoice for the amount to charge and run the PayPal charge via PayPal API. The invoice will then be updated to reflect whether it was paid successfully or not.
+
+Pros:
+
+- Stripe continues to serve as an accurate source of truth regarding a customers subscription.
+- All the logic that determines active subscriptions and entitlements using Stripe can be used as-is.
+- Multiple product support works as-is.
+- Support representatives can see accurate payment history in Stripe for customers paying via PayPal.
+- Support can credit a PayPal customer, just as they can for a credit card customer.
+- Can extend in a similar manner to track a correlating IAP for a user.
+- Can handle switching payment method from credit card to PayPal reference ID.
+
+Cons:
+
+- Additional metadata in Stripe can start pushing us closer to metadata limits on objects.
+- Sensitive processing logic that runs PayPal transactions which we need to ensure work correctly.
+- We still could end up storing PayPal reference ID's in a separate db table to ensure we track
+  prior PayPal billing.
+
+### PayPal Subscriptions linked/tracked by FxA
+
+In this approach, we would create products/plans in PayPal, and link them back to the Stripe equivilant by updating Stripe metadata on the products/plans. FxA would also gain an additional database table that maps PayPal subscription id's to the Fx account. All the logic in FxA that looks up a users subscriptions would be updated to check the PayPal linking table to determine what subscriptions the user has active.
+
+PayPal would handle running the charges on a recurring basis per the PayPal plan that is setup. FxA would periodically check the subscription's via PayPal API to ensure they're still actively being paid (PayPal has fairly lenient handling on whether payments occur) on the schedule we want.
+
+Pros:
+
+- PayPal is responsible for charging the customer and managing the subscription lifecycle.
+- Can extend in a similar manner to track a correlating IAP for a user.
+
+Cons:
+
+- All the logic that determines active subscriptions and entitlements needs additional
+  logic/abstractions to query the database table and map active PayPal subscriptions to the
+  Stripe subscription.
+- Ability to easily cancel/re-subscribe on management page is lost, as canceled PayPal subscriptions
```

Comment:
This seems to imply that we can easily resubscribe with the reference id.  Related question then: how and when does a reference id become invalid?

Reply:
It's more that since the subscription is invoice'd and tracked by Stripe, just as we do now, we can just mark the subscription as cancel_at_period_end. If they hit 're-subscribe', we just undo that mark. Cancelled PayPal subscriptions cannot be uncancelled in their system, a new subscription object entirely is needed, which muddies up when/how that needs to be done. In this case I wasn't referring to cancelling the reference ID.

To the related question, when a reference transaction is used, we get a flexible billing agreement that lets us run transactions without requiring further user approval. This billing agreement can be canceled by us, or by the user. The end-effect would be that our system would be in a similar state as when a user with a subscription has no valid cards on file. We would then have our subscription management page prompt the user to either put a new PayPal agreement on file, or a credit card for the next invoice cycle.